File: leetcode/209-minimum-size-subarray-sum/main.go
/*
Problem: Minimum Size Subarray Sum
Link: https://leetcode.com/problems/minimum-size-subarray-sum/

Solution: Use "Sliding Window"
*/
package main

import "math"

func minSubArrayLen(target int, nums []int) int {
	left := 0
	sum := 0
	result := math.MaxInt

	for right := 0; right < len(nums); right++ {
		sum += nums[right]

		for sum >= target {
			subLen := right - left + 1
			if subLen < result {
				result = subLen
			}
			sum -= nums[left]
			left++
		}
	}

	if result == math.MaxInt {
		return 0
	}
	return result
}

File: leetcode/209-minimum-size-subarray-sum/Main.kt
package main

fun main() {
  val solution = Solution()
  println(solution.minSubArrayLen(target = 7, nums = intArrayOf(2, 3, 1, 2, 4, 3)))
  // 2
}

class Solution {
  fun minSubArrayLen(target: Int, nums: IntArray): Int {
    var left = 0
    var right = 0
    var acc = 0
    var minLen = Int.MAX_VALUE

    while (right < nums.size) {
      acc += nums[right]

      while (acc >= target) {
        val len = right - left + 1
        minLen = if (len < minLen) len else minLen
        acc -= nums[left]
        left++
      }
      right++
    }

    return if (minLen == Int.MAX_VALUE) 0 else minLen
  }
}